training/tf/tfprocess.py

```diff
@@ -434,38 +441,57 @@ def construct_net(self, planes):
 
         return h_fc1, h_fc3
 
-    def save_swa_network(self, steps, path, leela_path, prev_swa, output_nets):
-        n = output_nets // self.swa_c
-        if self.swa_max_n != None:
-            n = min(n, self.swa_max_n)
-
-        swa_path = path + "-swa-" + str(n + 1) + "-" + str(steps) + ".txt"
+    def snap_save(self):
+        # Save a snapshot of all the variables in the current graph.
+        if not hasattr(self, 'save_op'):
+            save_ops = []
+            rest_ops = []
+            for var in self.weights:
+                if isinstance(var, str):
+                    var = tf.get_default_graph().get_tensor_by_name(var)
+                name = var.name.split(':')[0]
+                v = tf.Variable(var, name='save/'+name, trainable=False)
+                save_ops.append(tf.assign(v, var))
+                rest_ops.append(tf.assign(var, v))
+            self.save_op = tf.group(*save_ops)
+            self.restore_op = tf.group(*rest_ops)
+        self.session.run(self.save_op)
+
+    def snap_restore(self):
+        # Restore variables in the current graph from the snapshot.
+        self.session.run(self.restore_op)
+
+    def save_swa_network(self, steps, path, leela_path):
+        # Sample 1 in self.swa_c of the networks. Compute in this way so
+        # that it's safe to change the value of self.swa_c
+        rem = self.session.run(tf.assign_add(self.swa_skip, -1))
+        if rem > 0:
+            return
+        self.swa_skip.load(self.swa_c, self.session)
+
+        # Add the current weight vars to the running average.
+        num = self.session.run(self.swa_accum_op)
 
-        if not os.path.isfile(prev_swa):
-            # Average of one network is the network itself
-            copyfile(leela_path, swa_path)
-        else:
-            if self.swa_recalc_bn:
-                swa([prev_swa, leela_path], 'swa_temp.txt', weights=[n, 1])
-            else:
-                swa([prev_swa, leela_path], swa_path, weights=[n, 1])
+        if self.swa_max_n != None:
+            num = min(num, self.swa_max_n)
+            self.swa_count.load(float(num), self.session)
 
-        if n > 0 and self.swa_recalc_bn:
-            # Load SWA weights for batch norm recalculation
-            version, blocks, channels, weights = read_weights('swa_temp.txt')
-            self.replace_weights(weights)
+        swa_path = path + "-swa-" + str(num) + "-" + str(steps) + ".txt"
 
+        # save the current network.
+        self.snap_save()
+        # Copy the swa weights into the current network.
+        self.session.run(self.swa_load_op)
+        if self.swa_recalc_bn:
             print("Recalculating SWA batch normalization")
-            for _ in range(200):
+            self.update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
```

Comment:
Does this need to be redefined? It's already defined once in init_net?

Reply:
thanks :) fixed.